realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,279 @@
+
+        return (E) refAndCount.localRealm.get();
+    }
+
+    /**
+     * Releases a given {@link Realm} or {@link DynamicRealm} from cache. The instance will be closed by this function
+     * if there is no more local reference to this Realm instance in current Thread.
+     *
+     * @param realm Realm instance to be released from cache.
+     */
+    static synchronized void release(BaseRealm realm) {
+        String canonicalPath = realm.getPath();
+        RealmCache cache = cachesMap.get(canonicalPath);
+        Integer refCount = null;
+        RefAndCount refAndCount = null;
+
+        if (cache != null) {
+            refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realm.getClass()));
+            refCount = refAndCount.localCount.get();
+        }
+        if (refCount == null) {
+            refCount = 0;
+        }
+
+        if (refCount <= 0) {
+            RealmLog.w("Realm " + canonicalPath + " has been closed already.");
+            return;
+        }
+
+        // Decrease the local counter.
+        refCount -= 1;
+
+        if (refCount == 0) {
+            // The last instance in this thread.
+            // Clear local ref & counter
+            refAndCount.localCount.set(null);
+            refAndCount.localRealm.set(null);
+
+            // Clear global counter
+            refAndCount.globalCount--;
+            if (refAndCount.globalCount < 0) {
+                // Should never happen.
+                throw new IllegalStateException("Global reference counter of Realm" + canonicalPath +
+                        " got corrupted.");
+            }
+
+            // Clear the column indices cache if needed
+            if (realm instanceof Realm && refAndCount.globalCount == 0) {
+                // All typed Realm instances of this file are cleared from cache
+                cache.typedColumnIndices = null;
+            }
+
+            int totalRefCount = 0;
+            for (RealmCacheType type : RealmCacheType.values()) {
+                totalRefCount += cache.refAndCountMap.get(type).globalCount;
+            }
+            // No more instance of typed Realm and dynamic Realm. Remove the configuration from cache.
+            if (totalRefCount == 0) {
+                cachesMap.remove(canonicalPath);
+            }
+        } else {
+            refAndCount.localCount.set(refCount);
+        }
+
+        // No more local reference to this Realm in current thread.
+        if (refCount == 0) {
+            realm.doClose();
+        }
+    }
+
+    /**
+     * Make sure that the new configuration doesn't clash with any existing configurations for the
+     * Realm.
+     *
+     * @throws IllegalArgumentException If the new configuration isn't valid.
+     */
+    private void validateConfiguration(RealmConfiguration newConfiguration) {
+        // != takes no time, check it first
+        if (newConfiguration == configuration || newConfiguration.hashCode() == configuration.hashCode()) {
```

Comment:
Difference of `deleteRealmIfMigrationNeeded` and `migration` are not checked below.

We need to remove the comment `// Should never get here` or to add more checks for those differences.